include/xtensor-blas/xlinalg.hpp

```diff
@@ -1514,6 +1514,172 @@ namespace linalg
         }
         return res;
     }
+
+    /**
+     * @brief Compute tensor dot product along specified axes for arrays
+     *
+     * Compute the sum of products along the last \em naxes axes of a and first
+     * \em naxes axes of b.
+     *
+     * @param xa input array
+     * @param xb input array
+     * @param naxes the number of axes to sum over
+     * @return resulting array
+     */
+    template <class T, class O>
+    auto tensordot(const xexpression<T>& xa, const xexpression<O>& xb, std::size_t naxes = 2)
+    {
+        using value_type = std::common_type_t<typename T::value_type, typename O::value_type>;
+        using result_type = std::conditional_t<T::static_layout == O::static_layout &&
+                                               (T::static_layout != layout_type::dynamic && T::static_layout != layout_type::any),
+                                               xarray<value_type, T::static_layout>,
+                                               xarray<value_type, XTENSOR_DEFAULT_LAYOUT>>;
+
+        result_type result;
+        const auto& a = xa.derived_cast();
+        const auto& b = xb.derived_cast();
+        if (naxes == 0)
+        {
+            // special case tensor outer product product
+            std::vector<std::size_t> result_shape(a.dimension() + b.dimension());
+            std::size_t j = 0;
+            for (std::size_t i = 0; i < a.dimension(); ++i)
+            {
+                result_shape[j++] = a.shape()[i];
+            }
+
+            for (std::size_t i = 0; i < b.dimension(); ++i)
+            {
+                result_shape[j++] = b.shape()[i];
+            }
+            // flatten a/b
+            auto vec_a = xt::ravel<T::static_layout>(a);
+            auto vec_b = xt::ravel<O::static_layout>(b);
+            // take the outer product of the two vectors
+            result = outer(vec_a, vec_b);
+            // reshape the result
+            result.reshape(result_shape);
+        }
+        else
+        {
+          // Sum of products over last n axes of A and the first n axis of b
+            XTENSOR_ASSERT(a.dimension() >= axes);
+            XTENSOR_ASSERT(b.dimension() >= axes);
+
+            auto as_it = a.shape().begin() + (a.dimension() - naxes);
+            auto bs_it = b.shape().begin();
+            std::size_t sum_len = 1;
+            for (std::size_t i = 0; i < naxes; ++i)
+            {
+                auto a_val = *as_it;
+                auto b_val = *bs_it;
+                // check for axes size match
+                if (a_val != b_val)
+                {
+                    throw std::runtime_error("Shape mismatch for sum\n");
+                }
+                else
+                {
+                    sum_len *= a_val;
+                }
+                ++as_it;
+                ++bs_it;
+            }
+            std::vector<std::size_t> result_shape;
+            auto r_shape_inserter = std::back_inserter(result_shape);
```

Comment:
Maybe it would be slightly easier to follow the code if just using `push_back` on the `result_shape` instead?

Reply:
Agreed. I will change that.